### PR TITLE
Converter for SelectV2 operation.

### PR DESCRIPTION
**Pull request by drivanov**
Implementation of converter for [`ops::SelectV2`](https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/select-v2) operation.

### CODE REVIEW

File: tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc:59]: Please add a period to the end of the comment.

**Review by bixia1 — COMMENTED** [review]
Nice! Please also squash the commits into one.

**Review by bixia1 — APPROVED** [review]

**Comment by drivanov** [comment]
@bixia1: I am still working on improving this PR. In particular, I implement the ability to use broadcast forms for the `cond`, `then`, and `else` tensors. My last check-in already contains most of this implementation, but I plan to add a few more tests. For instance, currently, the case when all three tensors have **different** broadcastable shapes is not covered.

File: tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc]: Done

**Review by bixia1 — APPROVED** [review]

**Comment by drivanov** [comment]
@bixia1: I am done with all tests for that PR.

File: tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc
Comment (tfeher) [inline tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc:68]: This means that the condition input cannot be a weight, right? Why do we have this limitation?

File: tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc
Comment (tfeher) [inline tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc:64]: Is this variable used anywhere?

File: tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc
Comment (tfeher) [inline tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc:79]: Will this work with the following input shapes [2, 1, 1], [1, 2, 1], [1, 1, 2]?

- I believe you will need to reconcile broadcasted_dims[1] and proadcasted_dims[2], and then adjust broadcasted_dims[0] accordingly.
- Alternatively you could add on optional third input dim for GetTrtBroadcastShape.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (tfeher) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:9472]: Please add  {3, {2, 1, 1}}, {3, {1, 1, 2}}, {3, {1, 2, 1}} to the test.

**Review by tfeher — CHANGES_REQUESTED** [review]
Hi Andrei, thanks for the PR, overall it looks great! I think there is one issue with broadcasting that shall be addressed and tested.

File: tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc:68]: @tfeher : The point is that the condition input of `addSelect(...)` must be of type `nvinfer1::DataType::kBOOL`,  otherwise, we get an error:
```
2022-07-22 ... : E tensorflow/ ...  /my_select-selectv2:SELECT: condition tensor must have boolean type
```
But, unfortunately, today we cannot pass boolean as weight due to an error:
``` 
2022-07-22 ...: E tensorflow/...: (Unnamed Layer* 0) [Constant]: invalid weights type of Bool
```
That's why I introduced this restriction.

File: tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc:64]: Actually, this variable was used in my previous implementation of `ConvertSelectV2::Convert()`, which did NOT use the  `GetTrtBroadcastShape(...)` method. Perhaps we will need it again, when and if we will be able to pass `cond` as weight.

Remember, `GetTrtBroadcastShape(...)` works only when at least one of its parameters is a tensor.

For now, I will remove the calculation of `all_tensors_`.

File: tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc:79]: Unfortunately, it is impossible to reconcile `broadcasted_dims[1]` and `broadcasted_dims[2]`, when both `then` and `else` parameters are passed as weights. On the other hand, today we allow passing `cond` only as a tensor, I am using 
```
GetTrtBroadcastShape(inputs.at(0), inputs.at(i), ...
```

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:9472]: Done.

**Review by bixia1 — APPROVED** [review]

**Comment by bixia1** [comment]
@drivanov can you take care of this, it say cla/google failure.

File: tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc:72]: Then we should make it possible to reconcile when both `then` and `else` are weights.

File: tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc
Comment (tfeher) [reply tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc]: Thanks for the explanation! This is a TRT limitation. Please add a note to state that TRT does not allow boolean weights.

**Review by tfeher — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc]: DONE

File: tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc:72]: Yes, it is possible and we have [a subtest](https://github.com/tensorflow/tensorflow/blob/690d347af047876756c492024dd502f7497826ee/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc#L9316-L9324) for this, which checks all possible combinations of the passed parameters.

**Comment by drivanov** [comment]
@bixia1 : I'm done with this PR. Please review. Thanks!

**Review by bixia1 — APPROVED** [review]

File: tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/ops/selectv2.cc:87]: variable not used.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:9292]: error: use of undeclared identifier 'nextTensorWeigtConfiguration'

**Review by bixia1 — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:9292]: @bixia1 : I am not sure why `nextTensorWeigtConfiguration` is undeclared. This function is defined [in these lines](https://github.com/tensorflow/tensorflow/blob/690d347af047876756c492024dd502f7497826ee/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc#L3953-L3958) of `convert_nodes_test.cc`. Similar function was implemented in one of the previous PRs, but it was defined as a lambda function. So I think it might be some sort of merge issue.
Anyway, I'll fix the typo and change the name of this function from  `nextTensorWeigtConfiguration` to  `nextTensorWeightConfiguration`.

**Review by bixia1 — APPROVED** [review]

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc
Comment (bixia1) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc]: The function definition is protected via #if IS_TRT_VERSION_GE(8, 2, 0, 0)
but the use of it is not.

Comment (drivanov) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc]: @bixia1 : I'm sorry I didn't notice it. Corrected.

**Review by bixia1 — APPROVED** [review]

**Review by bixia1 — CHANGES_REQUESTED** [review]
emailed error log to Jonathan (for forwarding to the author).

**Review by bixia1 — APPROVED** [review]

**Comment by drivanov** [comment]
I fixed this issue by deleting the `ConvertSelectV`2 class and the corresponding test for TRT < 8.2.
There is no point in having them because this converter expects `cond` to be passed as a boolean tensor, which can only be created for TRT >= 8.2.

**Comment by drivanov** [comment]
@bixia1: Please, don't merge that one into master. I have resolved the formal merge conflicts appeared after the merge of [PR#56942](https://github.com/tensorflow/tensorflow/pull/56942), but I need to fix something.

**Comment by drivanov** [comment]
@bixia1: I fixed all merge conflicts. This PR is ready to merge.

**Review by bixia1 — APPROVED** [review]

**Comment by rthadur** [comment]
Can you please rebase your branch to head, Thank you.

**Comment by drivanov** [comment]
REBASED

**Review by bixia1 — APPROVED** [review]

**Pull request merged**
